### PR TITLE
You can now share your bike on the dashboard and see the share holders! 

### PR DESCRIPTION
**Pull request by sanderDijkxhoorn**

### CODE REVIEW

**Review by mjarkk — CHANGES_REQUESTED** [review]
This looks like a nice start of something new :)

Beside the points above i also have some ui feedback
image

1. It would be nice if there is a title above this section like `Share bike`
2. A bit of padding between the share and the disconnect button would be nice

File: components/ShareBike.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                        list = "tickmarks"
+                        onChange = { 
+                            e => setShareinfo(v => ({
+                                ...v, duration: Number(e.target.value)
+                            }))
+                        }
+                        value = { shareinfo.duration }
+                    />
+
+                    <Button type="submit" style={ ButtonStyling }>
```
Comment (mjarkk) [inline components/ShareBike.tsx:70]: the buttonStyleing is so small i think you can inline that value here
```suggestion
                    <Button type="submit" style={{margin: "1rem"}}>
```

File: lib/api.ts [hunk 2]
```diff
@@ -52,6 +52,35 @@ export class Api {
...
+                'Api-Key': API_KEY,
+                'Authorization': 'Bearer ' + this.credentials.token,
+                'Content-Type': 'application/json'
+            },
+            body: JSON.stringify(shareinfo)
+        })
+
+        if (req.status === 200) {
+            const res = await req.json()
+            // setSuccessModal(true) - TODO
```
Comment (mjarkk) [inline lib/api.ts:67]: This comment can be removed as this class should not create modals, that's up to the component wherefrom this method was called.

File: lib/api.ts [hunk 3]
```diff
@@ -52,6 +52,35 @@ export class Api {
...
+            return res
```
Comment (mjarkk) [inline lib/api.ts]: I think we can inline this:
```suggestion
            return await req.json()
```

**Review by mjarkk — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -8,6 +8,17 @@ A web app for changing the speed limit of your VanMoof S3 and X3.
...
+- Change your speed limit to JP (24 km/h), EU (27 km/h), US (32 km/h) or 😎 (37 km/h)
+- Set your power level, this also unlocks a new power level 5.
+- A sound board with the following sounds:
+
+Short | 🔘 Click | 🧨 Error | 👍 Pling | 🤔 Cling clong | 🔔 Bell | 🔔 Normal bike bell | 🎉 Bell Tada | 😚 Whistle | 🚢 BOAT | ⚡️ Wuup | 🫤 Success but error | 
+--- | --- | --- | --- |--- |--- |--- |--- |--- |--- |--- |---
+Long | 🔋 Charding noise.. | 🚨 Alarm | 🚨 Alarm stage 2 | 🔋 Charging.. | 🆕 Updating.. | 🎉 Update complete | 💥 Make wired noises
+- You can share your bike to someone using their email
+
+
```
Comment (mjarkk) [inline README.md:21]: I would normally like this to be in another PR as it's not relevant to this one tough for now lets go forward with it :)

File: components/ShareBike.tsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    const [error, setError] = useState<string | undefined>(undefined)
+    const [shareinfo, setShareinfo] = useState({
+        email: "",
+        bikeId: bike.id,
+        role: "user",
+        duration: 86400 // 86400 = 1 day but in seconds.
+    })
+
+    const onSubmit = async (event: FormEvent<HTMLFormElement>) => {
+        event.preventDefault()
```
Comment (mjarkk) [inline components/ShareBike.tsx:18]: I think we should clear the error message on resubmission
```js
setError(undefined)
```

File: components/ShareBike.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        try {
+            const result = await api.createBikeSharingInvitation(shareinfo)
+            if(result.result) {
+                // TODO: success popup
```
Comment (mjarkk) [inline components/ShareBike.tsx:22]: A popup would indeed be nice but to much for this pr

Tough i think a user should at least get some feedback the bike has been shared..

I think we should add a state value `success` (`boolean`) that is set to true if it succeeded and if it succeeded show a message similar to the error with a success message

**Review by mjarkk — CHANGES_REQUESTED** [review]

File: components/CurrentShares.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        } catch (e) {
+            setError(`${e}`)
+        }
+    }
+
+    return (
+        <div className='shareBike'>
+            <h3>Currently shared with</h3>
+
+            <div className='listContainer'>
```
Comment (mjarkk) [inline components/CurrentShares.tsx:46]: Maybe we should only show this element if `successModal` is `true`

File: components/CurrentShares.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            <form onSubmit={LoadInvites}>
+                <Button
+                    type="submit"
+                    style={{ margin: "1rem" }
+                    }>
+                    Get shared list
+                </Button>
+            </form>
+
+            <FormSuccess status={successModal} message={"Successfully obtained share holders."} />
```
Comment (mjarkk) [inline components/CurrentShares.tsx:82]: If we change what i suggested above (only showing the list of shares / showing a message you have no-one invited yet) we will have visual feedback if the data is loaded correctly thus making the `<FormSuccess />` unnecessary

File: components/ShareBike.tsx [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                        className = "inputCustom"
+                        type = "number"
+                        min = "1"
+                        onChange = { 
+                            e => setShareinfo(v => ({
+                                ...v, duration: Number(e.target.value) * 86400
+                            }))
+                        }
+                        value = { shareinfo.duration / 86400 }
+                    />
```
Comment (mjarkk) [inline components/ShareBike.tsx:87]: Can we change this to using `<ShareDurationSlider .. />`

**Review by mjarkk — CHANGES_REQUESTED** [review]

File: components/CurrentShares.tsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+            </div>
+
+            <form onSubmit={LoadInvites}>
+                <Button
+                    type="submit"
+                    style={{ margin: "1rem" }
+                    }>
+                    Get shared list
+                </Button>
+            </form>
```
[the remark this answers is not in the document]
Comment (sanderDijkxhoorn) [reply components/CurrentShares.tsx:80]: Could be useful if you want to refresh the share holders list

File: components/ShareBike.tsx (hunk 9 above)
Comment (sanderDijkxhoorn) [reply components/ShareBike.tsx:87]: If it works definitely :)

File: components/CurrentShares.tsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+            setError(`${e}`)
+        }
+    }
+
+    return (
+        <div className='shareBike'>
+            <h3>Currently shared with</h3>
+
+            {successModal
+                ? <div className='listContainer'>
```
[the remark this answers is not in the document]
Comment (sanderDijkxhoorn) [reply components/CurrentShares.tsx:73]: I agree, have you made this already or should I create this myself?

[the remark this answers is not in the document]
Comment (mjarkk) [reply components/CurrentShares.tsx:73]: Nope this is just a mockup

File: next.config.js [hunk 12]
```diff
@@ -5,8 +5,20 @@ const nextConfig = {
...
+          source: '/api/api_vanmoof-api_com/:path',
+          destination: 'https://api.vanmoof-api.com/v8/:path',
+        },
+        {
+          source: '/api/api_vanmoof-api_com/getBikeSharingInvitationsForBike/:path',
+          destination: 'https://api.vanmoof-api.com/v8/getBikeSharingInvitationsForBike/:path',
+        },
+        {
+          source: '/api/api_vanmoof-api_com/revokeBikeSharingInvitation/:path',
+          destination: 'https://api.vanmoof-api.com/v8/revokeBikeSharingInvitation/:path',
```
[the remark this answers is not in the document]
Comment (sanderDijkxhoorn) [reply next.config.js:21]: Not sure why but I couldn't seem to get it to work using just that.
